src/utilities/activityPubHash.ts

```diff
@@ -0,0 +1,33 @@
+import { keccak256 } from "js-sha3";
+import { KeccakHash } from "../types/hash";
+
+const sortJSON = (obj: Record<string, unknown>): Record<string, unknown> => {
+  const result: Record<string, unknown> = {};
+
+  Object.keys(obj)
+    .sort()
+    .forEach((key) => {
+      if (typeof obj[key] == "object") {
+        result[key] = sortJSON(obj[key] as Record<string, unknown>);
+      } else {
+        result[key] = obj[key];
+      }
+    });
+
+  return result;
+};
+
+/**
+ * activityPubHash() provides a simple way to hash activityPub objects while
+ * guaranteeing that identical objects with different key orders still return
+ * the same hash. The underlying hash method used is
+ * [Keccak256](https://en.wikipedia.org/wiki/SHA-3).
+ *
+ * @param data  The activityPub object to hashes
+ * @returns     A hexadecimal string containing the Keccak hash
+ */
+export const activityPubHash = (data: Record<string, unknown>): KeccakHash => {
+  const sortedData = sortJSON(data as Record<string, unknown>);
```

Comment:
I was imagining from the spec that the hash was happening upon the raw file, not the json. If we are going to change this to the json, we should update the spec.

Also sortJSON only sorts objects recursively, but not if there is an array of objects. Also I don't think JSON.parse or JSON.stringify guarantee order.

Reply:
`JSON.stringify` stringifies in the order the keys are present in the object, and I don't think we want to sort arrays since they have a semantic order. Sorting the keys in records here is just to guarantee an identical hash.

Reply:
JSON stringify does, but it isn't required to by the JSON spec.

Also this doesn't recursively sort objects in arrays:
```json
{
  attachments: [ {}, {} ]
}
```

Well ordered JSON is just not easy to guarantee due to the JSON spec.

Reply:
So, it turns out this edge case gets sorted too, because arrays are technically subclasses of objects in javascript. An array like `const x = [1, 2, 3]` will get sorted by this method because `typeof x == "object"` will return true and `Object.keys(x)` will return `[0, 1, 2]`.

Reply:
Hmm.... Let's continue this conversation in the spec side. We can leave it for now, but we should write the spec and discuss it there.

Reply:
Added a spec story: https://www.pivotaltracker.com/story/show/177681451